### ISSUE TITLE
[BE] better testing for subclasses + compile

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#114398]
We have plenty of tests for tensor subclasses with compile today [here](https://github.com/pytorch/pytorch/blob/main/test/functorch/test_aotdispatch.py#L3121)

But there are some areas that we could test better. A few that we don't test at all today include:

* torch.compile two different subclasses, e.g. `f(SubclassA(torch.Tensor), SubclassB(torch.Tensor))`

* torch.compile two *nested* subclasses, e.g. `f(SubclassA(SubclassB(torch.Tensor)))`

* torch.compile a function that uses both tensor subclass inputs, and an ambient mode:
```
with DispatchModeA():
    out = f(subclassBinput)
```

cc @Chillee @ezyang @zou3519 @albanD @samdow @msaroufim @wconstab @anijain2305

**Comment by Chillee** [comment]
I think we just need way more tests of composability, including things like tensor subclasses with higher order ops.

**Comment by bdhirsh** [comment]
Agreed - related issue: https://github.com/pytorch/pytorch/issues/114411

**Comment by hameerabbasi** [comment]
I've added a few tests in #158742
